posts/2018-12-21-Procedural-Macros-in-Rust-2018.md

```diff
@@ -0,0 +1,363 @@
+
+Each of the three types of procedural macros are [defined in a slightly different
+fashion][proc-ref], and here we'll single out attribute macros. First we'll flag
+`Cargo.toml`:
+
+```toml
+[lib]
+proc-macro = true
+```
+
+and then in `src/lib.rs` we can write our macro:
+
+```rust
+extern crate proc_macro;
+use proc_macro::TokenStream;
+
+#[proc_macro_attribute]
+pub fn hello(attr: TokenStream, item: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+We can then write some unit tests in `tests/smoke.rs`:
+
+```rust
+#[my_crate::hello]
+fn wrapped_function() {}
+
+#[test]
+fn works() {
+    wrapped_function();
+}
+```
+
+... and that's it! When we execute `cargo test` Cargo will compile our
+procedural macro. Afterwards it will compile our unit test which loads the macro
+at compile time, executing the `hello` function and compiling the resulting
+syntax.
+
+Right off the bat we can see a few important properties of procedural macros:
+
+* The input/output is this fancy `TokenStream` type we'll talk about more in a
+  bit
+* We're *executing aribtrary code* at compile time, which means we can do just
+  about anything!
+* Procedural macros are incorporated with the module system, meaning no more
+  janky `#[macro_use]`
+
+Before we take a look at implementing a procedural macro, let's first dive into
+some of these points.
+
+### Macros and the module system
+
+First stabilized in [Rust 1.30] \(noticing a trend with 1.15?\) macros are now
+integrated with the module system in Rust. This mainly means that you no longer
+need the clunky `#[macro_use]` attribute when importing macros! Instead of this:
+
+```rust
+#[macro_use]
+extern crate log;
+
+fn main() {
+    debug!("hello, ");
+    info!("world!");
+}
+```
+
+you can do:
+
+```rust
+use log::info;
+
+fn main() {
+    log::debug!("hello, ");
+    info!("world!");
+}
+```
+
+Integration with the module system solves one of the most confusing parts about
+macros historically. They're now imported and namespaced just as you would any
+other item in Rust!
+
+The benefits are not only limited to bang-style `macro_rules` macros, as you can
+now transform code that looks like this:
+
+```rust
+#[macro_use]
+extern crate serde_derive;
+
+#[derive(Deserialize)]
+struct Foo {
+    // ...
+}
+```
+
+into
+
+```rust
+use serde::Deserialize;
+
+#[derive(Deserialize)]
+struct Foo {
+    // ...
+}
+```
+
+and you don't even need to explicitly depend on `serde_derive` in `Cargo.toml`!
+
+### What's inside a `TokenStream`?
+
+This mysterious `TokenStream` type comes from the [compiler-provided
+`proc_macro` crate][pm]. When it was first added all you could do with a
+[`TokenStream`] was call `to_string()` or `parse()` one back from a string.
+As of Rust 2018, the [`TokenStream`] type provides a much larger stable API.
+
+As its name might imply, a `TokenStream` can be thought of as a stream of
+tokens, or a list of tokens. Each token is called a [`TokenTree`] and represents
+one of four kinds of tokens:
```

Comment:
I would replace the first half of this para with "A `TokenStream` is effectively "just" an iterator over `TokenTree`." from below. And perhaps move "All syntax in Rust falls into one of the above four categories." here too